limiter.go

```diff
@@ -0,0 +1,39 @@
+package hoglet
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sync/semaphore"
+)
+
+func newLimiter(origFactory observerFactory, limit int64, block bool) observerFactory {
+	sem := semaphore.NewWeighted(limit)
+
+	wrappedFactory := func(ctx context.Context) (observer, error) {
+		o, err := origFactory(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return observableCall(func(b bool) {
+			o.observe(b)
+			sem.Release(1)
+		}), nil
+	}
+
+	if block {
+		return func(ctx context.Context) (observer, error) {
+			if err := sem.Acquire(ctx, 1); err != nil {
+				return nil, fmt.Errorf("%w: %w", ErrWaitingForSlot, err)
+			}
+			return wrappedFactory(ctx)
+		}
+	}
+	return func(ctx context.Context) (observer, error) {
+		if sem.TryAcquire(1) {
```

Comment:
if block is false and sem.TryAcquire(1) fails, no permit is acquired from the sem, but we still release it, when the observer is called

Reply:
but if `TryAcquire` fails we return an error, so we never get the observer? What am I missing?

Reply:
oh, there's a lint error here: we can ditch the `else` 🙈

Reply:
ok, lint error gone; 

still think the TryAcquire is ok, but maybe I'm overlooking something?

@obitech WDYT?

Reply:
The way I understand it is if `TryAcquire` fails the lock can't be acquired and we return `ErrConcurrencyLimitReached` and the semaphore is left unchanged. So we should be good?

> TryAcquire acquires the semaphore with a weight of n without blocking. On success, returns true. On failure, returns false and leaves the semaphore unchanged.